dll/win32/shell32/folders/CFSFolder.cpp

```diff
@@ -129,67 +129,117 @@ HRESULT GetCLSIDForFileType(PCUIDLIST_RELATIVE pidl, LPCWSTR KeyName, CLSID* pcl
+
+static BOOL
+getShellClassInfo(LPCWSTR Entry, LPWSTR pszValue, DWORD cchValueLen, LPCWSTR IniFile)
+{
+    return GetPrivateProfileStringW(s_shellClassInfo, Entry, NULL, pszValue, cchValueLen, IniFile);
+}
+
+static HRESULT
+getIconLocationForFolder(IShellFolder * psf, LPCITEMIDLIST pidl, UINT uFlags,
+                         LPWSTR szIconFile, UINT cchMax, int *piIndex, UINT *pwFlags)
 {
-    static const WCHAR shellClassInfo[] = { '.', 'S', 'h', 'e', 'l', 'l', 'C', 'l', 'a', 's', 's', 'I', 'n', 'f', 'o', 0 };
     static const WCHAR iconFile[] = { 'I', 'c', 'o', 'n', 'F', 'i', 'l', 'e', 0 };
     static const WCHAR clsid[] = { 'C', 'L', 'S', 'I', 'D', 0 };
     static const WCHAR clsid2[] = { 'C', 'L', 'S', 'I', 'D', '2', 0 };
     static const WCHAR iconIndex[] = { 'I', 'c', 'o', 'n', 'I', 'n', 'd', 'e', 'x', 0 };
+    static const WCHAR iconResource[] = { 'I', 'c', 'o', 'n', 'R', 'e', 's', 'o', 'u', 'r', 'c', 'e', 0 };
     static const WCHAR wszDesktopIni[] = { 'd','e','s','k','t','o','p','.','i','n','i',0 };
-    int icon_idx;
 
-    if (!(uFlags & GIL_DEFAULTICON) && (_ILGetFileAttributes(ILFindLastID(pidl), NULL, 0) & (FILE_ATTRIBUTE_SYSTEM | FILE_ATTRIBUTE_READONLY)) != 0 )
+    do
     {
+        if (uFlags & GIL_DEFAULTICON)
+            break;
+
+        // read-only or system folder?
+        const DWORD dwFileAttrs = _ILGetFileAttributes(ILFindLastID(pidl), NULL, 0);
+        if ((dwFileAttrs & (FILE_ATTRIBUTE_SYSTEM | FILE_ATTRIBUTE_READONLY)) == 0)
+            break;
+
+        // get path
         WCHAR wszFolderPath[MAX_PATH];
+        SHGetPathFromIDListW(pidl, wszFolderPath);
+        if (!PathIsDirectoryW(wszFolderPath))
+            break;
 
-        if (!ILGetDisplayNameExW(psf, pidl, wszFolderPath, 0))
-            return FALSE;
+        // build a full path of ini file
+        WCHAR wszIniFullPath[MAX_PATH];
+        StringCchCopyW(wszIniFullPath, _countof(wszIniFullPath), wszFolderPath);
+        PathAppendW(wszIniFullPath, wszDesktopIni);
 
-        PathAppendW(wszFolderPath, wszDesktopIni);
+        // get current folder
+        WCHAR wszCurDir[MAX_PATH];
+        GetCurrentDirectoryW(_countof(wszCurDir), wszCurDir);
 
-        if (PathFileExistsW(wszFolderPath))
+        WCHAR wszValue[MAX_PATH], wszTemp[MAX_PATH];
+        if (getShellClassInfo(iconFile, wszValue, _countof(wszValue), wszIniFullPath))
         {
-            WCHAR wszPath[MAX_PATH];
-            WCHAR wszCLSIDValue[CHARS_IN_GUID];
+            // wszValue --> wszTemp
+            ExpandEnvironmentStringsW(wszValue, wszTemp, _countof(wszTemp));
 
-            if (GetPrivateProfileStringW(shellClassInfo, iconFile, NULL, wszPath, MAX_PATH, wszFolderPath))
+            if (SetCurrentDirectoryW(wszFolderPath))
             {
-                ExpandEnvironmentStringsW(wszPath, szIconFile, cchMax);
+                // wszTemp --> szIconFile
+                GetFullPathNameW(wszTemp, cchMax, szIconFile, NULL);
```

Comment:
Is it possible to avoid the usage of the thread-unsafe SetCurrentDirectory function?
I was thinking about using some FindFirstFile... but maybe other solutions exist?

Reply:
PathCombine or something in shlwapi shall be used. I will try to implement it tomorrow.

Reply:
I used PathIsRelativeW and PathAppendW. Done.